isis/tests/FunctionalTestsJigsaw.cpp

```diff
@@ -0,0 +1,246 @@
+#include <iostream>
+
+#include <QFile>
+#include <QHash>
+#include <QSet>
+#include <QString>
+#include <QStringList>
+#include <QTextStream>
+
+#include "jigsaw.h"
+#include "Angle.h"
+#include "ControlNet.h"
+#include "ControlPoint.h"
+#include "ControlMeasure.h"
+#include "Distance.h"
+#include "FileName.h"
+#include "Fixtures.h"
+#include "Latitude.h"
+#include "Longitude.h"
+#include "SurfacePoint.h"
+#include "TestUtilities.h"
+#include "gmock/gmock.h"
+#include "UserInterface.h"
+#include "CSVReader.h"
+
+using namespace Isis;
+
+static QString APP_XML = FileName("$ISISROOT/bin/xml/jigsaw.xml").expanded();
+
+// Check to see if a QString contains only numeric values.
+bool isNumeric(QString str){
+  QRegExp re("-*\\d*.*\\d*");  // a digit (\d), zero or more times (*)
+  return re.exactMatch(str);
+}
+
+// Testing help function which will compare csv lines
+void compareCsvLine(CSVReader::CSVAxis csvLine, QString headerStr, int initialIndex=0) { 
+  QStringList compareMe = headerStr.split(",");
+  for (int i=initialIndex; i<compareMe.size(); i++) {
+    if (isNumeric(compareMe[i].trimmed())) {
+      EXPECT_NEAR(csvLine[i].toDouble(), compareMe[i].toDouble(), 0.000001);
+    }
+    else{
+      EXPECT_EQ(QString(csvLine[i]).toStdString(), compareMe[i].toStdString()); 
+    }
+  }
+};
+
+
+TEST_F(ApolloNetwork, FunctionalTestJigsawFirst) {
+  QString controlNetFile = tempDir.path()+"/apollo.net";
+  QString inputNetFile =  tempDir.path()+"/apolloNet.net";
+  QVector<QString> args1 = {"radius=yes",
+                            "errorpropagation=yes",
+                            "spsolve=position",
+                            "spacecraft_position_sigma=1000.0",
+                            "camsolve=angles",
+                            "twist=yes",
+                            "camera_angles_sigma=2.",
+                            "update=no",
+                            "bundleout_txt=no",
+                            "cnet="+tempDir.path()+"/apolloNet.net",
+                            "fromlist="+tempDir.path() + "/cubes.lis",
+                            "onet="+tempDir.path()+"/apollo_out.net",
+  "file_prefix="+tempDir.path()+"/"};
+
+  UserInterface ui1(APP_XML, args1);
+
+  jigsaw(ui1);
+
+  // Test points.csv, images.csv, residuals.csv 
+
+  QString pointsOutput = tempDir.path() + "/bundleout_points.csv";
+  QString imagesOutput = tempDir.path() + "/bundleout_images.csv";
+  QString residualsOutput = tempDir.path() + "/residuals.csv";
+  ControlNet outputNet(tempDir.path()+"/apollo_out.net");
+
+  // Check for the correct header output format and csv file structure for the points.csv file
+  CSVReader::CSVAxis csvLine;
+  CSVReader header = CSVReader(pointsOutput,
+                               false, 0, ',', false, true);
+
+  int numColumns = header.columns();
```

Comment:
This process is definitely better for comparing headers than what I was doing. That being said, as jigsaw is kinda all over the place (in the sense of it relies on a number of objects). Try to check what parts of the CSV isn't already tested by one of these objects as part of a unittest (I think bundleobservation and bundleresults do a lot of the work? there is also bundleUtilities/unittest.cpp that does a lot of testing).

Reply:
Good point. 

I did some grepping, and it looks like the header values for the any of the csv files aren't being tested anywhere, so I think it makes sense to leave the header comparisons in for each of the output csv files, too, and just test them in this _one_ jigsaw test. I don't think the headers need to be tested for any other jigsaw tests, since they're covered here.

The other csv line comparisons just "spot-check" a few points to make sure their exact values match the expected values in the current output, though I have no particular preference for doing this as opposed to taking the statics approach you took. I'm tempted to leave all of this in. What do you think?